File: Solutions/cpp/trafficsignal/Signal.cpp
#include "Singal.h"

Signal::Signal(const std::string& id, int green, int yellow, int red)
    : signalId(id), greenDuration(green), yellowDuration(yellow), redDuration(red), currentSignal(SignalType::RED), isWorking_(true) {}

Signal::~Signal() {}

std::string Signal::getSignalId() const {
    return signalId;
}

SignalType Signal::getCurrentSignal() const {
    return currentSignal;
}

int Signal::getRemainingDuration() const {
    return remaingDuration;
}

bool Signal::isWorking() const {
    return isWorking_;
}

void Signal::setSignal(SignalType signal) {
    currentSignal = signal;
    switch (signal) {
        case SignalType::GREEN:
            currentSignal = SignalType::GREEN;
            remaingDuration = greenDuration;
            break;
        case SignalType::YELLOW:
            currentSignal = SignalType::YELLOW;
            remaingDuration = yellowDuration;
            break;
        case SignalType::RED:
            currentSignal = SignalType::RED;
            remaingDuration = redDuration;
            break;
        default:
            remaingDuration = 0;
            break;
    }
}

void Signal::updateSignal(int timeElapsed) {
    if (!isWorking_) return;
    if (remaingDuration > timeElapsed) {
        remaingDuration -= timeElapsed;
    } else {
        remaingDuration = 0;
        switchSignal();
    }
}

void Signal::setDuration(int green, int yellow, int red) {
    greenDuration = green;
    yellowDuration = yellow;
    redDuration = red;
    // 若当前信号为某色，重置剩余时间
    switch (currentSignal) {
        case SignalType::GREEN:
            remaingDuration = greenDuration;
            break;
        case SignalType::YELLOW:
            remaingDuration = yellowDuration;
            break;
        case SignalType::RED:
            remaingDuration = redDuration;
            break;
        default:
            remaingDuration = 0;
            break;
    }
}

void Signal::setWorking(bool working) {
    isWorking_ = working;
}

void Signal::reset() {
    currentSignal = SignalType::RED;
    remaingDuration = redDuration;
    isWorking_ = true;
}

void Signal::switchSignal() {
    if (!isWorking_) return;
    switch (currentSignal) {
        case SignalType::RED:
            currentSignal = SignalType::GREEN;
            remaingDuration = greenDuration;
            break;
        case SignalType::GREEN:
            currentSignal = SignalType::YELLOW;
            remaingDuration = yellowDuration;
            break;
        case SignalType::YELLOW:
            currentSignal = SignalType::RED;
            remaingDuration = redDuration;
            break;
        default:
            currentSignal = SignalType::RED;
            remaingDuration = redDuration;
            break;
    }
}

void Signal::printStatus() const {
    std::string signalStr;
    switch (currentSignal) {
        case SignalType::RED:
            signalStr = "RED";
            break;
        case SignalType::GREEN:
            signalStr = "GREEN";
            break;
        case SignalType::YELLOW:
            signalStr = "YELLOW";
            break;
        default:
            signalStr = "UNKNOWN";
            break;
    }
    std::cout << "Signal[" << signalId << "] Status: " << signalStr
              << ", Remaining: " << remaingDuration
              << ", Working: " << (isWorking_ ? "Yes" : "No") << std::endl;
}

File: Solutions/golang/trafficsignal/signal.go
package main

import "fmt"

type SignalType int

const (
	RED SignalType = iota
	GREEN
	YELLOW
	UNKNOWN
)

func (s SignalType) String() string {
	switch s {
	case RED:
		return "RED"
	case GREEN:
		return "GREEN"
	case YELLOW:
		return "YELLOW"
	default:
		return "UNKNOWN"
	}
}

type Signal struct {
	signalId          string
	currentSignal     SignalType
	remainingDuration int
	greenDuration     int
	yellowDuration    int
	redDuration       int
	isWorking         bool
	isEmergency       bool
}

func NewSignal(signalId string, greenDuration, yellowDuration, redDuration int) *Signal {
	return &Signal{
		signalId:          signalId,
		currentSignal:     RED,
		remainingDuration: redDuration,
		greenDuration:     greenDuration,
		yellowDuration:    yellowDuration,
		redDuration:       redDuration,
		isWorking:         false,
		isEmergency:       false,
	}
}

func (s *Signal) GetSignalId() string {
	return s.signalId
}

func (s *Signal) GetCurrentSignal() SignalType {
	return s.currentSignal
}

func (s *Signal) GetRemainingDuration() int {
	return s.remainingDuration
}

func (s *Signal) GetGreenDuration() int {
	return s.greenDuration
}

func (s *Signal) GetYellowDuration() int {
	return s.yellowDuration
}

func (s *Signal) GetRedDuration() int {
	return s.redDuration
}

func (s *Signal) GetIsWorking() bool {
	return s.isWorking
}

func (s *Signal) SetSignal(signal SignalType) {
	s.currentSignal = signal
}

func (s *Signal) UpdateSignal(timeElapsed int) {
	if !s.isWorking {
		return
	}
	if s.remainingDuration > timeElapsed {
		s.remainingDuration -= timeElapsed
	} else {
		s.remainingDuration = 0
		s.SwitchSignal()
	}
}

func (s *Signal) SetDuration(greenDuration, yellowDuration, redDuration int) {
	s.greenDuration = greenDuration
	s.yellowDuration = yellowDuration
	s.redDuration = redDuration
	s.remainingDuration = redDuration
}

func (s *Signal) SetWorking(working bool) {
	s.isWorking = working
}

func (s *Signal) Reset() {
	s.currentSignal = RED
	s.remainingDuration = s.redDuration
	s.isWorking = true
	s.isEmergency = false
}

func (s *Signal) SwitchSignal() {
	if !s.isWorking {
		return
	}
	switch s.currentSignal {
	case RED:
		s.currentSignal = GREEN
		s.remainingDuration = s.greenDuration
	case GREEN:
		s.currentSignal = YELLOW
		s.remainingDuration = s.yellowDuration
	case YELLOW:
		s.currentSignal = RED
		s.remainingDuration = s.redDuration
	}
}

func (s *Signal) PrintStatus() {
	fmt.Printf("  Signal[%s] Status: %s, Remaining: %d sec, Working: %t, Emergency: %t\n",
		s.signalId, s.currentSignal, s.remainingDuration, s.isWorking, s.isEmergency)
}

// HandleEmergency sets the signal to GREEN immediately for emergency vehicles
func (s *Signal) HandleEmergency() {
	if !s.isWorking {
		return
	}
	s.isEmergency = true
	s.currentSignal = GREEN
	s.remainingDuration = s.greenDuration
}

// ClearEmergency clears the emergency state
func (s *Signal) ClearEmergency() {
	s.isEmergency = false
}
